fv3core/decorators.py

```diff
@@ -90,120 +92,118 @@ def wrapped(state, *args, **kwargs):
         """Return the build_info created when compiling the stencil."""
-        return self._build_info
 
-    def __call__(self, *args, **kwargs):
-        return self.stencil_object(*args, **kwargs)
+        self.compute_domain: Optional[Tuple[Int3, Int3]] = None
+        """The last compute domain (origin, domain) that this was called using."""
 
+        self.times_called: int = 0
+        """Number of times this stencil has been called."""
 
-def _ensure_global_flags_not_specified_in_kwargs(stencil_kwargs):
-    flag_errmsg = (
-        "The {} flag should be set in "
-        + __name__
-        + " instead of as an argument to stencil"
-    )
-    for flag in ("rebuild", "backend"):
-        if flag in stencil_kwargs:
-            raise ValueError(flag_errmsg.format(flag))
+        self.externals: Dict[str, Any] = kwargs.pop("externals", {})
+        """External dictionary used for stencil generation."""
 
+        self.backend_kwargs: Dict[str, Any] = kwargs
+        """Remainder of the arguments are assumed to be gt4py compiler backend options."""
 
-def gtstencil(definition=None, **stencil_kwargs) -> Callable[..., None]:
-    _ensure_global_flags_not_specified_in_kwargs(stencil_kwargs)
+    @property
+    def built(self) -> bool:
+        """Returns True if the stencil is already built (if it has been called at least once)."""
+        return self.stencil_object is not None
 
-    def decorator(func) -> Callable[..., None]:
-        stencils = {}
-        times_called = 0
-
-        def get_origin(func, call_args, call_kwargs):
-            sig = inspect.signature(func)
-            first_name = next(iter(sig.parameters))
-
-            if len(call_args) == 0:
-                first_storage = call_kwargs[first_name]
-            else:
-                first_storage = call_args[0]
-
-            if not isinstance(first_storage, gt4py.storage.storage.Storage):
-                raise TypeError(
-                    "First stencil argument should be a gt4py.storage.storage.Storage."
-                )
-
-            origin: Sequence[int, ...]
-            origin_arg = call_kwargs.get("origin", None)
-            if isinstance(origin_arg, collections.Mapping):
-                if first_name in origin_arg:
-                    origin = origin_arg[first_name]
-                else:
-                    origin = first_storage.default_origin
-            elif isinstance(origin_arg, collections.Sequence):
-                origin = origin_arg
-            elif origin_arg is None:
-                origin = first_storage.default_origin
-            else:
-                raise TypeError(
-                    f"Type of the origin argument was {type(origin_arg)}. Expected Sequence or Mapping."
-                )
-
-            return origin
+    @property
+    def def_ir(self) -> Optional[dict]:
+        """Current stencil definition IR if built, else None."""
+        return self.build_info.def_ir
```

Comment:
adding this for debugging purposes?

Reply:
Yeah, that future data flow analysis!

cc: @eddie-c-davis